### ISSUE TITLE
Add stein variational gradient descent as sampling possibility, for increased performance

### ISSUE THREAD

**Issue by jdehning** [Priesemann-Group/covid19_inference#6]
In PyMC3, stein variational gradient is already implemented, but it has to be tested how well it works/how biased it is for small number of particles. In addition, the optimal type of optimizer and learning rate has to found out. Eventually it could be tried to reparametrize the model, to get a simpler posterior

**Comment by ac-schneider** [comment]
http://www.cs.utexas.edu/~lqiang/PDF/svgd_aabi2016.pdf

**Comment by ac-schneider** [comment]
In principle one can use the ASVGD inference, however, one might have to play with the priors a bit to help the algorithm converge. Also, it is not yet recommended to use. What helped in my case to make it converge more robustly:

- mean prior for lambda_0 of 0.6.
- prior for I_0 of around 50.
- adam with learning rate 0.02
- 300 steps at least .. which takes around 6 minutes in a google colab notebook

**Comment by tarbaig** [comment]
General comment: If you are hell bent on using variational methods, consider switching to pyro,
imho the variational methods are more mature there.
https://pyro.ai/

Any particular reason why variational stuff should be better? People use it to train bayesian neural networks and with those number of parameters sampling becomes unfeasible. But given the few parameters of the current model that should not be a problem.

**Comment by jdehning** [comment]
We are not hell bent on it. My thought was that variational methods could eventually be parallelized, when we will look at the level of Landkreise.
But that I find that interesting, intrinsically are Monte-Carlo methods scaling worse than variational methods with the number of parameters? My take was that the advanced methods like Hamiltonian MC scale pretty well with the number of parameters. But my knowledge is pretty superficial.

**Comment by jdehning** [comment]
One this issue, no one is actively working on. So if someone want to have a look...
Things learned so far: 

- the pymc3.ASVGD needs very high temperature, around 2 to give more or less the correct posterior, one the 1 dimensional problem.
- It takes time until ASVGD and SVGD converges, mainly probably because of wide distribution of I_begin. Learning rates around 0.01 seems to work.

Next steps would be use pymc3.SVGD with about 100 particles and try to apply it to the example_bundeslaender, to see whether one gets faster to approximate good results. There one can also test whether theano uses multiprocessing.

**Comment by dmarinere** [comment]
So basically how can people help @jdehning 

**A different question** 
how do we infer these rates for different countries?
Are they universal?
